### PR TITLE
EVM-483 Relayer Micro Service

### PR DESCRIPTION
**Pull request by igorcrevar**
# Description

Account Abstraction - Relayer Micro-service: HTTP endpoint: (RFC-195)

Relayer micro service can be started with new command:
`polygon-edge aarelayer --addr 127.0.0.1:8198 --data-dir ./test-chain-1`

Note that secrets must be generated previously in `./test-chain-1` directory

Rest server exposes two endpoints:
- `POST` `/v1/sendTransaction`
    the body should be json encoded `AATransaction` from the `aa_types.go`
- `GET` `/v1/getTransactionReceipt/[uuid]`

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [X] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [X] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [ ] I have tested this code manually

### CODE REVIEW

File: command/aarelayer/service/aa_config.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	"encoding/json"
+
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+//go:embed config/*
+var configFile embed.FS
+
+type AAConfig struct {
+	Whitelist []string `json:"whitelist"`
```
Comment (vcastellm) [inline command/aarelayer/service/aa_config.go]: `AllowList`

File: command/aarelayer/service/aa_config.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	Blacklist []string `json:"blacklist"`
```
Comment (vcastellm) [inline command/aarelayer/service/aa_config.go]: `DenyList` 🤷

File: command/aarelayer/service/aa_types.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+)
+
+var marshalArenaPool fastrlp.ArenaPool
+
+// string mapping
+const (
+	StatusPending   string = "pending"   // The AA transaction is on the Pool
+	StatusQueued           = "queued"    // The AA transaction is waiting to be mined.
+	StatusCompleted        = "completed" // The `AA transaction` was mined in a block.
+	StatusFinished         = "finished"  // AA transaction` failed during the process.
```
Comment (vcastellm) [inline command/aarelayer/service/aa_types.go]: finished = failed?

File: command/aarelayer/service/config/config.json [hunk 4]
```diff
@@ -0,0 +1,3 @@
+{
+    "whitelist": null
+}
```
Comment (vcastellm) [inline command/aarelayer/service/config/config.json]: Missing LF at the end of the file, but why is this included in the PR?

File: command/aarelayer/params.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import (
+	"fmt"
+
+	"github.com/0xPolygon/polygon-edge/command/helper"
+)
+
+const (
+	addrFlag = "addr"
+
+	defaultPort = 8198
```
Comment (vcastellm) [inline command/aarelayer/params.go:17]: Why this port selection?

File: command/aarelayer/service/aa_pool.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
+package service
+
+import (
+	"sync"
+)
+
+type AAPool interface {
```
Comment (vcastellm) [inline command/aarelayer/service/aa_pool.go:8]: Can we add some comments in this file?

File: command/aarelayer/service/aa_state.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
+package service
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/google/uuid"
+)
+
+type AATxState interface {
```
Comment (vcastellm) [inline command/aarelayer/service/aa_state.go:11]: Also some comments to this file.

File: command/aarelayer/service/aa_server.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+// GetTransactionReceipt handles the /v1/getTransactionReceipt/{uuid} endpoint
+func (s *AARelayerRestServer) getTransactionReceipt(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodGet {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+
+		return
+	}
+
+	uuid := r.URL.Path[len("/v1/getTransactionReceipt/"):]
```
Comment (vcastellm) [inline command/aarelayer/service/aa_server.go]: Would be great to move this to a constant

File: command/aarelayer/service/aa_server.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+		return
+	}
+
+	if err := s.verification.Validate(&tx); err != nil {
+		writeMessage(w, http.StatusBadRequest, err.Error())
+
+		return
+	}
+
+	// put to DB
```
Comment (vcastellm) [inline command/aarelayer/service/aa_server.go]:
```suggestion
	// store in state
```

File: command/aarelayer/service/aa_pool.go [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+	Put(string, *AATransaction)
+	Pull() *AAPoolTransaction
```
Comment (vcastellm) [inline command/aarelayer/service/aa_pool.go]: I would aim for the more standard `Push/Pop` even though I see it's consistent with mashalArena and keccakPool, so consider this nitpicking.

In any case, is the Pull being used at all?

File: command/aarelayer/service/aa_config.go (hunk 1 above)
Comment (igorcrevar) [reply command/aarelayer/service/aa_config.go]: I have copied name from the rfc https://www.notion.so/polygontechnology/RFC-194-e14733e47dc548b4b12384965367ce3f (`whitelisted`). Maybe we should change naming there also?

File: command/aarelayer/service/aa_types.go (hunk 3 above)
Comment (igorcrevar) [reply command/aarelayer/service/aa_types.go]: Just copied naming from rfc https://www.notion.so/polygontechnology/RFC-194-e14733e47dc548b4b12384965367ce3f. Maybe we should change there also?

File: command/aarelayer/params.go (hunk 5 above)
Comment (igorcrevar) [reply command/aarelayer/params.go:17]: We can use any other port. Suggestions?

File: command/aarelayer/service/aa_pool.go (hunk 10 above)
Comment (igorcrevar) [reply command/aarelayer/service/aa_pool.go]: It will be used. This PR is just base for all the others which will come latter.

But I will rename those to `Push/Pop`

File: command/aarelayer/service/aa_pool.go (hunk 6 above)
Comment (igorcrevar) [reply command/aarelayer/service/aa_pool.go:8]: This is just base, correct version of pool will be implemented via separated task

File: command/aarelayer/service/config/config.json (hunk 4 above)
Comment (igorcrevar) [reply command/aarelayer/service/config/config.json]: I thought it is just small functionality so i wanted to push it through this pr :). Is it ok if I just add LF at the end?

File: command/aarelayer/service/aa_pool.go (hunk 10 above)
Comment (igorcrevar) [reply command/aarelayer/service/aa_pool.go]: fixed https://github.com/0xPolygon/polygon-edge/pull/1252/commits/f8527534f60da34f423360164a5bfd3a2cbf29d9

File: command/aarelayer/service/aa_server.go (hunk 9 above)
Comment (igorcrevar) [reply command/aarelayer/service/aa_server.go]: fixed https://github.com/0xPolygon/polygon-edge/pull/1252/commits/f8527534f60da34f423360164a5bfd3a2cbf29d9

File: command/aarelayer/service/aa_server.go (hunk 8 above)
Comment (igorcrevar) [reply command/aarelayer/service/aa_server.go]: Fixed https://github.com/0xPolygon/polygon-edge/pull/1252/commits/f8527534f60da34f423360164a5bfd3a2cbf29d9

File: command/aarelayer/service/aa_state.go (hunk 7 above)
Comment (igorcrevar) [reply command/aarelayer/service/aa_state.go:11]: Fixed https://github.com/0xPolygon/polygon-edge/pull/1252/commits/f8527534f60da34f423360164a5bfd3a2cbf29d9

File: command/aarelayer/service/aa_pool.go (hunk 6 above)
Comment (igorcrevar) [reply command/aarelayer/service/aa_pool.go:8]: Added https://github.com/0xPolygon/polygon-edge/pull/1252/commits/f8527534f60da34f423360164a5bfd3a2cbf29d9

File: command/aarelayer/params.go (hunk 5 above)
Comment (igorcrevar) [reply command/aarelayer/params.go:17]: Changed to `8081` (`80` and `8080` are standard web server ports, I think we should avoid them)  https://github.com/0xPolygon/polygon-edge/pull/1252/commits/f8527534f60da34f423360164a5bfd3a2cbf29d9

File: command/aarelayer/service/config/config.json (hunk 4 above)
Comment (igorcrevar) [reply command/aarelayer/service/config/config.json]: `LF` on the end https://github.com/0xPolygon/polygon-edge/pull/1252/commits/f8527534f60da34f423360164a5bfd3a2cbf29d9

File: command/aarelayer/service/aa_config.go (hunk 2 above)
Comment (igorcrevar) [reply command/aarelayer/service/aa_config.go]: If we agree to change `Whitelist` to `Blacklist`, I will change this also

File: command/aarelayer/service/aa_types.go (hunk 3 above)
Comment (igorcrevar) [reply command/aarelayer/service/aa_types.go]: ...but yes, it is semantically more correct if we change `StatusFinished` to `StatusFailed`, in my opinion you are right

File: command/aarelayer/service/aa_config.go (hunk 1 above)
Comment (vcastellm) [reply command/aarelayer/service/aa_config.go]: I'd say yes

File: command/aarelayer/service/aa_config.go (hunk 2 above)
Comment (vcastellm) [reply command/aarelayer/service/aa_config.go]: It should be changed TBH

File: command/aarelayer/params.go (hunk 5 above)
Comment (vcastellm) [reply command/aarelayer/params.go:17]: Sure, and 8198 is ok! I just wanted to know if you checked something like https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers 8198 look good, 8081 has official reservation

File: command/aarelayer/service/aa_types.go (hunk 3 above)
Comment (igorcrevar) [reply command/aarelayer/service/aa_types.go]: Changed https://github.com/0xPolygon/polygon-edge/pull/1252/commits/396931ce5428a8d78bb2bba0bfa8da056edc7fcd

File: command/aarelayer/service/aa_config.go (hunk 2 above)
Comment (igorcrevar) [reply command/aarelayer/service/aa_config.go]: Changed https://github.com/0xPolygon/polygon-edge/pull/1252/commits/396931ce5428a8d78bb2bba0bfa8da056edc7fcd

File: command/aarelayer/service/aa_config.go (hunk 1 above)
Comment (igorcrevar) [reply command/aarelayer/service/aa_config.go]: Changed https://github.com/0xPolygon/polygon-edge/pull/1252/commits/396931ce5428a8d78bb2bba0bfa8da056edc7fcd

File: command/aarelayer/params.go (hunk 5 above)
Comment (igorcrevar) [reply command/aarelayer/params.go:17]: Reverted back to `8198` https://github.com/0xPolygon/polygon-edge/pull/1252/commits/396931ce5428a8d78bb2bba0bfa8da056edc7fcd
I did not check that list, but thanks, great resource!

File: command/aarelayer/service/aa_pool.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+// AAPool defines the interface for a pool of Account Abstraction (AA) transactions
+type AAPool interface {
+	// Push adds an AA transaction to the pool, associating it with the given account ID
+	Push(*AAStateTransaction)
+	// Pop removes the next transaction from the pool and returns a wrapper object containing the transaction
+	Pop() *AAStateTransaction
+	// Init initializes the pool with a set of existing AA transactions. Used on client startup
+	Init([]*AAStateTransaction)
```
Comment (goran-ethernal) [inline command/aarelayer/service/aa_pool.go:14]: Will we use this function anywhere in some newer PR? I don't see it used in this one.

Comment (igorcrevar) [reply command/aarelayer/service/aa_pool.go:14]: https://github.com/0xPolygon/polygon-edge/pull/1261 it will be used here

Comment (igorcrevar) [reply command/aarelayer/service/aa_pool.go:14]: And correct implementation will be in some other pr. I am currently working on pool

File: command/aarelayer/service/aa_config.go [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+	for _, v := range c.AllowList {
+		if addressStr == strings.TrimPrefix(v, "0x") {
+			return true
+		}
+	}
+
+	return false
+}
+
+func GetConfig() (*AAConfig, error) {
```
Comment (ferranbt) [inline command/aarelayer/service/aa_config.go]: I would prefer to declare this as:
```
func DefaultConfig() *AAConfig {
  return &AAConfig{}
}
```
following the standard in the repo. We can avoid returning an error in the function, the unmarshaling and the embed. I would still keep the embed in a test to check that we can decode a json file (user input) to the config.

**Review by ferranbt — CHANGES_REQUESTED** [review]

File: command/aarelayer/service/aa_config.go (hunk 12 above)
Comment (igorcrevar) [reply command/aarelayer/service/aa_config.go]: Fixed https://github.com/0xPolygon/polygon-edge/pull/1252/commits/55ed79ac314c5508b8b842c188aa5a42f7849f06

**Review by ferranbt — APPROVED** [review]

**Review by vcastellm — APPROVED** [review]

**Review by goran-ethernal — APPROVED** [review]

**Review by Stefan-Ethernal — APPROVED** [review]
:rocket:
